### ISSUE TITLE
Version support

### ISSUE THREAD

**Issue by maturmel** [MscrmTools/XrmToolBox#729]
Hi, this is more a question than an issue I think, I would like to confirm whether the latest XrmToolBox (2018) is able to connect to 2015 on-premises version of CRM? I see the XTB uses SDK v9 but my deployment of CRM is on SDK v7. There seem to be incompability when I execute some of my custom XTB plugins in the 2018 version, it fails to properly make the call to the CRM platform. It seems from the error message it is related to version compatibility.

So my underlying question is : am I able to run XTB 2018 with the v7 of SDK or do I have to stay on latest version that has v7 SDK (2015.9.21 probably)

Thanks

**Comment by maturmel** [comment]
My plugin uses a feature that is available in 2015, has it works under XTB 2015.9.21. But, I put the same plugin DLL in plugins folder of XTB 2018, it does not work.

Any operation that needs to do a write (or delete) to the CRM database returns an error.

"The formatter threw an exception while trying to deserialize the message: There was an error while trying to deserialize parameter http://schemas.microsoft.com/xrm/2011/Contracts/Services:request. The InnerException message was 'Error in line 1 position 14075. Element 'http://schemas.datacontract.org/2004/07/System.Collections.Generic:value' contains data from a type that maps to the name 'http://schemas.microsoft.com/xrm/7.1/Contracts:ConcurrencyBehavior'. The deserializer has no knowledge of any type that maps to this name. Consider changing the implementation of the ResolveName method on your DataContractResolver to return a non-null value for name 'ConcurrencyBehavior' and namespace 'http://schemas.microsoft.com/xrm/7.1/Contracts'.'.  Please see InnerException for more details."}

In case it would've helped, I've tried to regenerate the proxy code for my custom CRM entities with CrmSvcTool V9, it doesn't work because CrmSvcTool V9 cannot connect to CRM 2015 deployment. 

I have no clue where to look for with regards to the above .NET exception I am getting upon calling xrmContext.SaveChanges() in my code. The same code works fine in XTB 2015.9.21.

Regards

**Comment by MscrmTools** [comment]
It seems you try to use optimistic concurrency.
If so, this feature was introduced in CRM 2015 Online Update 1, so it is not available in CRM 2015 OnPremise.

**Comment by maturmel** [comment]
I appreciate your time and help.

However, I am not calling for optimistic concurrency explicitly in my plugin code anywhere. I am just using XrmServiceContext class as before.
As I already mentionned, exact same code works in XTB 2015.9.21. It cannot be using a feature that does not exist in that version. I make no change to the code, just copy same DLL in XTB 2018 folder and when running from XTB 2018 it returns the above error when calling SaveChanges()

Either something on the SDK or in XTB is not backward compatible with regards to concurrency feature introduced like you said in 2015 online update 1.

Thanks

**Comment by MscrmTools** [comment]
Would you be able to test your plugin with some intermediate version? Or is version 1.2015.9.21 is the latest where your code works?

**Comment by maturmel** [comment]
I tried following:

2015.9.21 -> loads my plugins and execute them correctly
2016.5.27 -> error when loading custom plugins (method get_Service() not implemented)
2016.6.29 -> same as above
2016.10.2 -> same as above
2017.1.6 -> unable to launch XTB, it said the path had locked files but I checked every file and there is none locked
2017.3.14 -> get_Service() error again
2017.5.15 -> same get_Service() error
2017.7.18 -> same get_Service() error
2017.10.19 -> loads my plugins properly. gives concurrency error on execution of SaveChanges SDK call

So it seems 2016.X until 2017.10.19 were not compatible with my plugin definition, I am missing implementation details it seems. I had not updated from 2015 until last fall when trying out the Export Excel plugin which is of use to me. I wish to regroup all my used plugins under same version of XTB for easier management. I have a lot of in-house plugins that do clean-up task that are specific to my CRM environment (uses custom entities).

So to simply answer your question, 2015.9.21 seems to be the latest working version, I could not test with 2016.X because of plugin definition changes, which 2017.10.19 and up fixes but then I get the platform concurrency error.

Do you have a suggestion of what to try next?

Thanks again for your time and help.

**Comment by MscrmTools** [comment]
I cannot tell more for now.

I’m in « vacation » because my second child born last week so I will return to work on March 16th.

Could you send me the plugin so that I can test it myself and try some debug session?

**Comment by maturmel** [comment]
Hi,

Congratulations to your family for the new born child :-) Take the deserved time off, by all means...

I would have a hard time sending you the plugin "as is" because it is targeted some custom entities that would not exist in other CRM deployments. However I can explain what the code does in a few simple steps:

1- runs a LINQ query to fetch some records in an entity (I set xrmContext.MergeOption = MergeOption.NoTracking;)
2- loops through the list of results
3- for each record, I call below code:

using (XrmServiceContext updateContext = new XrmServiceContext(this.service))
                            {
                                updateContext.Attach(record);
                                updateContext.DeleteObject(record);
                                updateContext.SaveChanges();
                            }

The error occurs on the SaveChanges() call.
Note: the above code was being called in a Parallel.ForEach loop but I tried with regular loop and it gives the same result.

Let me know if you need more information to try to replicate.

**Comment by daryllabar** [comment]
There are issues using Earlybound objects in plug-in calls.  MS does some work to try and find your EarlyBoundEntities for you, and sometimes may find someone else's in a different plugin.  Have you tried deleting all other plugins?  I'd also try reverting it to late bound calls and seeing what that does, especially deletes just because it's way easier using the IOrganizationService rather than the Organization service context attach and save changes methodology

**Comment by maturmel** [comment]
Hi Daryl,

In my case it cannot be mixing early bound entities because they are all contained within a DLL shared by all my custom XTB plugins. The entity names have a prefix which prevent them from being mixed up by other XTB-downloaded plugins. I also tried with removing all other plugins from the folder, leaving just mine, to be sure it is referencing correct early bound class.
Late bound could be an option for deletes like you said, but in case of updates in becomes more invasive changes in the code. I use early bound in all my crm code and would prefer to keep it like so for consistency and readability reasons.
For now I can live with 2015 XTB version to run my custom plugins. I can use the updated plugins from the store with the XTB 2018 shell and have both installed side by side on my workstation. It's not ideal but doesn't require any code change :)

I asked the question because I was curious to confirm whether the newer XTB was still backward compatible with older CRM deployments.

I will research on concurrency options that were added in later CRM versions to see how they can be mapped for the default behavior of my version, and see if I can do anything about that to resolve the error I posted above.

**Comment by maturmel** [comment]
I have also tried to reference v9 of SDK in my code, and XrmServiceContext class has property "ConcurrencyBehavior". I set it to "AlwaysOverwrite" has it would mimic the behavior of CRM 2015 on-premise, but despite that I still get the same error when calling SaveChanges(). It looks like my version of Dynamics is simply not compatible with SDK assemblies currently referenced by XTB. I guess I will have to remain with two XTB installations side-by-side to maintain 2015.9.21 version with working CRM 2015 plugins.
